core/src/main/scala/zio/jdbc/JdbcEncoder.scala

```diff
@@ -16,623 +16,271 @@
 package zio.jdbc
 
 import zio.Chunk
+import zio.jdbc.SqlFragment.{ Segment, values }
 import zio.schema.{ Schema, StandardType }
 
+import java.sql.{ PreparedStatement, Types }
+
 /**
  * A type class that describes the ability to convert a value of type `A` into
  * a fragment of SQL. This is useful for forming SQL insert statements.
  */
-trait JdbcEncoder[-A] {
+trait JdbcEncoder[-A] { self =>
   def encode(value: A): SqlFragment
 
-  final def contramap[B](f: B => A): JdbcEncoder[B] = value => encode(f(value))
+  /**
+   * Returns first index after encoder
+   */
+  def unsafeSetValue(ps: PreparedStatement, index: Int, value: A): Int
+
+  /**
+   * Returns first index after encoder
+   */
+  def unsafeSetNull(ps: PreparedStatement, index: Int): Int
+
+  final def contramap[B](f: B => A): JdbcEncoder[B] =
+    new JdbcEncoder[B] {
+      override def encode(value: B): SqlFragment =
+        self.encode(f(value))
+
+      override def unsafeSetValue(ps: PreparedStatement, index: Int, value: B): Int =
+        self.unsafeSetValue(ps, index, f(value))
+
+      override def unsafeSetNull(ps: PreparedStatement, index: Int): Int = self.unsafeSetNull(ps, index)
+
+      override def sql(a: B): String = "?"
+    }
+
+  def sql(value: A): String
```

Comment:
I think this should be implemented by calling `encode` and then rendering that `SqlFragment` to a SQL string.